File: 2021/task1/task1.js
const readline = require('readline').createInterface(process.stdin, process.stdout);

let accountsCount = null;
let managersCount = null;
const accountsList = [];
let totalAmount = 0;
let linesCount = 0;

const getTransactionsCount = (amount) => {
    return accountsList.filter(item => item >= amount).reduce((acc, item) => acc + Math.floor(item / amount), 0);
}

const solve = () => {
    accountsList.sort().reverse();

    if (totalAmount < managersCount) {
        return 0;
    }

    let minAmount = Math.floor(accountsList[0] / managersCount);
    let maxAmount = Math.floor(totalAmount / managersCount);
    let transactions = 0;
    let steps = 0;
    let midPoint = 0;

    while (true) {
        midPoint = (minAmount + maxAmount) / 2;
        transactions = getTransactionsCount(Math.round(midPoint));
        if (Math.abs(minAmount - maxAmount) <= 0.5 && transactions >= managersCount) {
            break;
        }

        if (transactions >= managersCount) {
            minAmount = midPoint
        } else {
            maxAmount = midPoint
        }
        steps++;
    }

    return Math.round(midPoint);
};

readline.on('line', (line) => {
    if (linesCount === 0) {
        [accountsCount, managersCount] = line.split(' ').map(item => Number(item));
    } else {
        totalAmount += Number(line);
        accountsList.push(Number(line));
    }
    if (linesCount++ === accountsCount) {
        console.log(solve());
        readline.close();
    }
}).on('close', () => process.exit(0));

File: 2021/task1/task1.py
N, M = [int(x) for x in input().split()]  # Читаем N и M в первой строке
C = [int(input()) for _ in range(N)]  # Читаем Cn в последующих строках


def solve(M, C):
    max_amount = sum(C) // M  # максимум - средняя сумма на менеджера
    # // - оператор деления с округлением в меньшую сторону до целого

    if max_amount == 0:  # если она равна 0 – можно выйти сразу
        return 0

    rounded_middle = 1
    min_amount = 1  # начинаем с 1, потому что 0 уже не может быть
    transactions = M

    # Бинарный поиск, второе условие компенсирует перебор
    while max_amount - min_amount >= 0.5 or transactions < M:
        # пробуем сумму между максимальным и минимальным
        middle = (min_amount + max_amount) / 2
        rounded_middle = round(middle)

        # вычисляем количество транзакций (целых премий) с такой суммой
        transactions = sum([x // rounded_middle for x in C if x >= rounded_middle])
        if transactions < M:
            max_amount = middle  # сужаем сверху при недоборе
        else:
            min_amount = middle  # сужаем снизу при переборе

    return rounded_middle

print(solve(M, C))